### PR TITLE
fix(menus): location metadata is not always a string

### PR DESCRIPTION
**Pull request by hypeJunction**
Multiple location metadata on a user flood the log with warnings

### CODE REVIEW

**Comment by hypeJunction** [comment]
Could someone restart the scrutinizer inspection? I don't have permissions

**Comment by juho-jaakkola** [comment]
Yeah, I'll restart them.

File: engine/lib/users.php [hunk 1]
```diff
@@ -564,8 +564,9 @@ function elgg_users_setup_entity_menu($hook, $type, $return, $params) {
 		$return = array(\ElggMenuItem::factory($options));
 	} else {
 		$return = array();
-		if (isset($entity->location)) {
-			$location = htmlspecialchars($entity->location, ENT_QUOTES, 'UTF-8', false);
+		if (isset($entity->location) && !empty($entity->location)) {
```
Comment (mrclay) [inline engine/lib/users.php]: Simpler and would save a bunch of function calls:
```php
$location = (array)$entity->location;
if ($location) {
    $location = $location[0];
    ...
}
```

Comment (hypeJunction) [inline engine/lib/users.php]: Trying to write a regression test, and it seems we need to filter the array and remove any non-string and empty values. Thoughts?

Comment (hypeJunction) [inline engine/lib/users.php]: @mrclay What's your call on this?

**Comment by juho-jaakkola** [comment]
Based on this we should make sure that the location always gets saved as a string: https://github.com/Elgg/Elgg/blob/master/engine/lib/upgrades/2013030600-1.8.13-update_user_location-8999eb8bf1bdd9a3.php#L21

**Comment by hypeJunction** [comment]
Well, that's against the metadata logic and I have plugins that imply that user can have multiple location metadata. The fix you are linking to wants to make sure we don't store location segments as separate metadata, but what about an object that can have multiple locations, e.g. Berlin, Germany and Helsinki, Finland

**Comment by juho-jaakkola** [comment]
We must find out why it was changed from array to string in 1.8.

**Comment by hypeJunction** [comment]
If I remember correctly, location was stored as tags.

**Comment by hypeJunction** [comment]
@brettp Do you have a recollection of the reasons?

**Comment by brettp** [comment]
I believe it was because tags used to be the primary way to search. I do remember each location section was tagged so you could click them and find other people with that location.

**Comment by mrclay** [comment]
I'm torn here. If a non-bundled plugin has created incompatible metadata structures under the same name as core, I'm not sure this is core's problem. Why shouldn't the plugin author write a script to clean up the data or otherwise handle it?

**Comment by mrclay** [comment]
...If, after debate in another ticket, we decide core should handle multiple locations, a PR should include everything necessary to input/output those, right? Correct me if I'm missing the point of this PR.

**Comment by hypeJunction** [comment]
I am fine with upgrading my plugins, however, we should document it clearly
which metadata is expected to be a string and short circuit if someone
tried to write multiple values. This holds true for location, lat/long,
status, etc.
The point if this ticket is not to show Array in the menu.

**Comment by hypeJunction** [comment]
Are we going to act on this or let it be?

**Comment by juho-jaakkola** [comment]
I don't have a strong opinion on the approach as long as we get rid of the "Array" string in the entity menu.

**Comment by hypeJunction** [comment]
I am just going to grab the first value

**Comment by mrclay** [comment]
@hypeJunction sound fine to me

**Comment by hypeJunction** [comment]
@mrclay I feel like ignoring your first comment on casting md as array. It would give an impression we want it to be an array. So, this is good to go in as is.

**Comment by mrclay** [comment]
If we don't want to give the impression that it's an array, we should use a comment, or as I suggested earlier, let plugin authors who store an array deal with the problem it creates.

If this problem is just to handle Elgg MD saving/timing bugs, then we need to broaden the scope and handle this lots of places we read MD, maybe using a new argument to getMetadata().

**Comment by beck24** [comment]
Reminds me of this: https://github.com/Elgg/Elgg/issues/4268

**Comment by mrclay** [comment]
Simplest solution here:
```php
$location = $entity->location;
if (is_string($location) && $location !== '') {
    // show it
```

**Comment by hypeJunction** [comment]
Agreed. Architecture corrupted by plugins will then be left to plugins to fix.

**Comment by juho-jaakkola** [comment]
Needs to be rebased on 1.11 branch

**Pull request closed without merging**